compiler/rustc_mir_build/src/check_tail_calls.rs

```diff
@@ -0,0 +1,387 @@
+use rustc_abi::ExternAbi;
+use rustc_errors::Applicability;
+use rustc_hir::LangItem;
+use rustc_hir::def::DefKind;
+use rustc_middle::span_bug;
+use rustc_middle::thir::visit::{self, Visitor};
+use rustc_middle::thir::{BodyTy, Expr, ExprId, ExprKind, Thir};
+use rustc_middle::ty::{self, Ty, TyCtxt};
+use rustc_span::def_id::{DefId, LocalDefId};
+use rustc_span::{DUMMY_SP, ErrorGuaranteed, Span};
+
+pub(crate) fn check_tail_calls(tcx: TyCtxt<'_>, def: LocalDefId) -> Result<(), ErrorGuaranteed> {
+    let (thir, expr) = tcx.thir_body(def)?;
+    let thir = &thir.borrow();
+
+    // If `thir` is empty, a type error occurred, skip this body.
+    if thir.exprs.is_empty() {
+        return Ok(());
+    }
+
+    let is_closure = matches!(tcx.def_kind(def), DefKind::Closure);
+    let caller_ty = tcx.type_of(def).skip_binder();
+
+    let mut visitor = TailCallCkVisitor {
+        tcx,
+        thir,
+        found_errors: Ok(()),
+        // FIXME(#132279): we're clearly in a body here.
```

Comment:
yes but it would define opaque types :3 

but maybe we want to use @lcnr's post borrowck analysis

Reply:
I'm not sure what options get us there, but I think this code should be accepted:
```rust
fn a() -> impl Sized {
    become b()
}

fn b() -> u8 { 0 }
```

which means that we should reveal opaques before comparing things 🤔

Reply:
well, so we probably want to reveal only the opaques local to the body

Reply:
i think this is good for now, and having a fixme maybe for deciding on behavior is good enough!